apps/sync/js/adapters/bookmarks.js

```diff
@@ -0,0 +1,325 @@
+  | keyword       | keyword       | keyword       |
+  | parentid      | parentid      | parentid      |
+  | parentName    | parentName    | parentName    |
+  | predecessorid | predecessorid | predecessorid |
+  | type          | type          | type          |
+  +---------------+---------------+---------------+
+
+  +---------------+---------------+---------------+
+  | folder        | livemark      | separator     |
+  +---------------+---------------+---------------+
+  |               | *siteUri      | *pos          |
+  |               | *feedUri      |               |
+  | title         | title         |               |
+  | parentid      | parentid      | parentid      |
+  | parentName    | parentName    | parentName    |
+  | predecessorid | predecessorid | predecessorid |
+  | type          | type          | type          |
+  +---------------+---------------+---------------+
+
+  [*] Special property for its type.
+  [**] microsummary is OBSOLETED. Please see here:
+       https://wiki.mozilla.org/Microsummaries
+
+  [5] https://docs.services.mozilla.com/sync/objectformats.html#bookmarks
+
+**/
+  _update(remoteRecords, lastModifiedTime) {
+    var bookmarks = [];
+    for (var i = 0; i < remoteRecords.length; i++) {
+      var payload = remoteRecords[i].payload;
+      if (payload.type === 'microsummary') {
+        console.warn('microsummary is OBSOLETED ', payload);
+        continue;
+      }
+      if (!Number.isInteger(remoteRecords[i].last_modified)) {
+        console.warn('Incorrect payload::last_modified? ', payload);
+        continue;
+      }
+      if (remoteRecords[i].last_modified <= lastModifiedTime) {
+        break;
+      }
+      if (payload.deleted) {
+        bookmarks.push({
+          deleted: true,
+          fxsyncId: payload.id
+        });
+        continue;
+      }
+      var typeWithUri = ['query', 'bookmark']
+          .some(value => value === payload.type);
+      if (typeWithUri && !payload.bmkUri) {
+        console.warn('Incorrect payload? ', payload);
+        continue;
+      }
+      bookmarks.push({
+        id: payload.type === 'bookmark' ? payload.bmkUri :
+          (payload.type + '|' + payload.id),
+        url: payload.bmkUri,
+        name: payload.title,
+        type: payload.type === 'bookmark' ? 'url' : 'others',
+        iconable: false,
+        icon: '',
+        fxsyncPayload: payload,
+        fxsyncId: payload.id
+      });
+    }
+
+    if (bookmarks.length === 0) {
+      return Promise.resolve(false);
+    }
+
+    return BookmarksHelper.updateBookmarks(bookmarks).then(() => {
+      var latestMtime = remoteRecords[0].last_modified;
+      return BookmarksHelper.setSyncedCollectionMtime(latestMtime);
+    }).then(() => {
+      // Always return false for a read-only operation.
+      return Promise.resolve(false);
+    });
+  },
+
+  update(remoteBookmarks) {
+    var mtime;
+    return LazyLoader.load(['shared/js/async_storage.js'])
```

Comment:
Instead of lazy loading `async_storage.js` before the function call, lazy load it within the function implementation. Also for the setter https://github.com/mozilla-b2g/gaia/pull/32131/files#diff-6dac2f6d58d68fbc11a10804f9946d4fR45